### PR TITLE
Allele-specific annotations and tests

### PR DESCRIPTION
**Pull request by akiezun**
this is the initial port of the Allele Specific annotation for HaplotypeCaller. It mostly focuses on the GVCF mode (ie outputs the 'raw' data). I have a branch in protected https://github.com/broadinstitute/gatk-protected/tree/ak_haplotypecaller_allele_specific_annotations that uses those and I verified that the annotations are correctly output and their values are much closer that before to those from GATK3.5. I did not port any code related to combining the annotations in GenotypeGVCFs or CombinedGVCFs etc. Also, no code for VariantAnnotator or UnifiedGenotyper was ported - gatk4 does not have those tools right now.

@droazen can you review? Sorry, this is a whole bunch of code and it's not the final version yet (in particular, little effort was put into redesigning the framework - that will wait until we have integration tests so we can keep the results stable while improving design and code)

We also need to add tickets to:
- turn dithering off/on in RankSum tests (it's always off for now to simplify testing)
- use AlleleSpecific annotations in the VCF mode
- (later) port code for combining annotations

### CODE REVIEW

**Comment by droazen** [comment]
@akiezun Can you check the original versions of the annotations classes in as a separate commit, in the same locations/names as the ported GATK4 versions?

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+    public Map<String, Object> annotateRawData(final ReferenceContext ref,
+                                               final VariantContext vc,
+                                               final Map<String, PerReadAlleleLikelihoodMap> perReadAlleleLikelihoodMap ) {
+
+        //for allele-specific annotations we only call from HC and we only use perReadAlleleLikelihoodMap
+        if ( perReadAlleleLikelihoodMap == null) {
+            return new HashMap<>();
+            // calculate the annotation from the stratified per read likelihood map
+            // stratifiedPerReadAllelelikelihoodMap can come from HaplotypeCaller call to VariantAnnotatorEngine
+        } else if (perReadAlleleLikelihoodMap != null) {
```
Comment (sputnikci) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java]: \[FindBugs\] WARNING: RCN: Redundant nullcheck of perReadAlleleLikelihoodMap, which is known to be non\-null in org\.broadinstitute\.hellbender\.tools\.walkers\.annotator\.AS\_StrandBiasTest\.annotateRawData\(ReferenceContext, VariantContext, Map\)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+
+    protected String makeRawAnnotationString(final List<Allele> vcAlleles, final Map<Allele, List<Integer>> perAlleleValues) {
+        String annotationString = "";
+        for (final Allele a : vcAlleles) {
+            if (!annotationString.isEmpty())
+                annotationString += printDelim;
+            List<Integer> alleleValues = perAlleleValues.get(a);
+            if (alleleValues == null)
+                alleleValues = ZERO_LIST;
+            annotationString += encode(alleleValues);
```
Comment (sputnikci) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java]: \[FindBugs\] INFO: SBSC: org\.broadinstitute\.hellbender\.tools\.walkers\.annotator\.AS\_StrandBiasTest\.makeRawAnnotationString\(List, Map\) concatenates strings using \+ in a loop

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        }
+        return annotationString;
+    }
+
+    protected String encode(List<Integer> alleleValues) {
+        String annotationString = "";
+        for (int j =0; j < alleleValues.size(); j++) {
+            annotationString += alleleValues.get(j);
+            if (j < alleleValues.size()-1)
+                annotationString += ",";
```
Comment (sputnikci) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java]: \[FindBugs\] INFO: SBSC: org\.broadinstitute\.hellbender\.tools\.walkers\.annotator\.AS\_StrandBiasTest\.encode\(List\) concatenates strings using \+ in a loop

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java (hunk 1 above)
Comment (sputnikci) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java]: \[FindBugs\] WARNING: RCN: Redundant nullcheck of perReadAlleleLikelihoodMap, which is known to be non\-null in org\.broadinstitute\.hellbender\.tools\.walkers\.annotator\.AS\_StrandBiasTest\.annotateRawData\(ReferenceContext, VariantContext, Map\)

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java (hunk 2 above)
Comment (sputnikci) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java]: \[FindBugs\] INFO: SBSC: org\.broadinstitute\.hellbender\.tools\.walkers\.annotator\.AS\_StrandBiasTest\.makeRawAnnotationString\(List, Map\) concatenates strings using \+ in a loop

File: src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java (hunk 3 above)
Comment (sputnikci) [inline src/main/java/org/broadinstitute/hellbender/tools/walkers/annotator/AS_StrandBiasTest.java]: \[FindBugs\] INFO: SBSC: org\.broadinstitute\.hellbender\.tools\.walkers\.annotator\.AS\_StrandBiasTest\.encode\(List\) concatenates strings using \+ in a loop

**Comment by droazen** [comment]
Waiting on original versions of annotation classes being checked in as a separate commit -- back to @akiezun. Easiest way is to create a new branch, copy GATK3 classes into their GATK4 locations/names, `git cherry-pick fbd1a402c0d73b3abb3f60a235ec70e978f88271`, and open a PR for the new branch.

**Comment by akiezun** [comment]
closing this in favor of #1840 (not deleting the branch for now, for backup)

**Pull request closed without merging**